sysboundary/inflow.cpp

```diff
@@ -0,0 +1,497 @@
+    FsGrid<array<Real, fsgrids::ehall::N_EHALL>, FS_STENCIL_WIDTH>& EHallGrid, cint i, cint j, cint k,
+    cuint component) {
+   std::array<Real, fsgrids::ehall::N_EHALL>* cp = EHallGrid.get(i, j, k);
+   switch (component) {
+   case 0:
+      cp->at(fsgrids::ehall::EXHALL_000_100) = 0.0;
+      cp->at(fsgrids::ehall::EXHALL_010_110) = 0.0;
+      cp->at(fsgrids::ehall::EXHALL_001_101) = 0.0;
+      cp->at(fsgrids::ehall::EXHALL_011_111) = 0.0;
+      break;
+   case 1:
+      cp->at(fsgrids::ehall::EYHALL_000_010) = 0.0;
+      cp->at(fsgrids::ehall::EYHALL_100_110) = 0.0;
+      cp->at(fsgrids::ehall::EYHALL_001_011) = 0.0;
+      cp->at(fsgrids::ehall::EYHALL_101_111) = 0.0;
+      break;
+   case 2:
+      cp->at(fsgrids::ehall::EZHALL_000_001) = 0.0;
+      cp->at(fsgrids::ehall::EZHALL_100_101) = 0.0;
+      cp->at(fsgrids::ehall::EZHALL_010_011) = 0.0;
+      cp->at(fsgrids::ehall::EZHALL_110_111) = 0.0;
+      break;
+   default:
+      abort_mpi("Invalid component", 1);
+   }
+}
+
+void Inflow::fieldSolverBoundaryCondGradPeElectricField(
+    FsGrid<std::array<Real, fsgrids::egradpe::N_EGRADPE>, 2>& EGradPeGrid, cint i, cint j, cint k, cuint component) {
+   EGradPeGrid.get(i, j, k)->at(fsgrids::egradpe::EXGRADPE + component) = 0.0;
+}
+
+void Inflow::fieldSolverBoundaryCondDerivatives(
+    FsGrid<std::array<Real, fsgrids::dperb::N_DPERB>, FS_STENCIL_WIDTH>& dPerBGrid,
+    FsGrid<std::array<Real, fsgrids::dmoments::N_DMOMENTS>, FS_STENCIL_WIDTH>& dMomentsGrid, cint i, cint j, cint k,
+    cuint RKCase, cuint component) {
+   this->setCellDerivativesToZero(dPerBGrid, dMomentsGrid, i, j, k, component);
+}
+
+void Inflow::fieldSolverBoundaryCondBVOLDerivatives(
+    FsGrid<std::array<Real, fsgrids::volfields::N_VOL>, FS_STENCIL_WIDTH>& volGrid, cint i, cint j, cint k,
+    cuint component) {
+   this->setCellBVOLDerivativesToZero(volGrid, i, j, k, component);
+}
+
+void Inflow::vlasovBoundaryCondition(const dccrg::Dccrg<SpatialCell, dccrg::Cartesian_Geometry>& mpiGrid,
+                                     const CellID& cellID, const uint popID, const bool doCalcMomentsV) {
+   if (dynamic) {
+      SpatialCell* cell = mpiGrid[cellID];
+
+      creal dx = cell->parameters[CellParams::DX];
+      creal dy = cell->parameters[CellParams::DY];
+      creal dz = cell->parameters[CellParams::DZ];
+      creal x = cell->parameters[CellParams::XCRD] + 0.5 * dx;
+      creal y = cell->parameters[CellParams::YCRD] + 0.5 * dy;
+      creal z = cell->parameters[CellParams::ZCRD] + 0.5 * dz;
+
+      bool isThisCellOnAFace[6];
+      determineFace(&isThisCellOnAFace[0], x, y, z, dx, dy, dz, true);
+
+      for (uint i = 0; i < 6; i++) {
+         if (facesToProcess[i] && isThisCellOnAFace[i]) {
+            copyCellData(&templateCells[i], cell, false, popID, true); // copy also vdf, _V
+            copyCellData(&templateCells[i], cell, true, popID, false); // don't copy vdf again but copy _R now
```

Comment:
Maybe I made unnecessary copies here. `copyCellData(&templateCells[i], cell, false, popID, doCalcMomentsV)` shall be enough. But even then, I copy VDFs twice, once post translation and once post acceleration. Should it be done only once, since the Vlasov BC is always applied at `P::t+0.5*P::dt`? @ykempf